Alex and Happy Part 1/article.md

```diff
@@ -159,27 +170,53 @@ Lexers are often implemented using Deterministic Finite Automata (FDAs), which y
+The dead ends represent an implicit _error state_ present in every DFA, which you can think of like a hidden state number 5, with all invalid transitions from every state to it.
+
+To simplify matters, on ambiguous transitions such as `i` and `[a-z]`, assume that the one with a single letter takes precedence.
 
-If we want to check the expression "int", we will follow this path: 0 → 1 → 2 → 4.
+For example, to check the expression "int", we will follow this path: 0 → 1 → 2 → 4.
 
 Now, we need to attribute meanings to each state so that this automaton can be useful, and we will do so like this:
 * 0: Error.
 * 1: Identifier.
 * 2: In.
 * 3: If.
 * 4: Identifier.
+* 5 (implicit): Error.
 
 Since we stopped at 4, the result is that "int" is an identifier, so we get a token that is `Identifier "int"`.
-The resulting token would be 'In' if we stopped at 2 instead (for the expression "in").
+The resulting token would be `In` if we stopped at 2 instead (for the expression "in").
 
-
-What if we consumed no input and stopped at state 0, or the input was something like "Foo123" where there aren't valid transitions?
+What if we consumed no input and stopped at state 0, or the input was something like "Foo123" where we'd reach the implicit state 5?
 In such cases, we halt and indicate a lexical error.
 
-We won't write our lexical analyzer from scratch. Instead, we will use Alex to create a list of tokens, which Happy will parse.
+As started previously, we won't write our lexical analyzer from scratch. Instead, we will use Alex to generate a lexer that creates a stream of tokens, which the Happy-generated parser will parse.
+
+### Regular expressions
+
+Alex uses regular expressions (regexes) to match patterns, we'll keep it pretty simple and explain some of them which we'll use.
+
+A syntax like `[0-9]` means that any digit character between `0` and `9` (inclusive) will be matched. Likewise, `[a-zA-Z]` means that any lowercase or uppercase character can be matched.
+Characters may be excluded as well, for example, `[^\"]` reads as "anything except a double quote mark".
+
+A `*` (called _Kleene star_) means that the expression can be matched zero or more times. Similarly, `+` means that it can be matched one or more times. For instance, `[0-9]+` can match `1234`. You may also find `?`, which means that an expression may match either zero or one time (i.e., it's optional).
+
+A dot (`.`) means "match anything but a newline". You can also use common escape codes like `\n` which will match a newline.
+
+An expression between strings, such as `">="`, means that it should only match exactly that string. An escaped character like `\?` can also match exactly that character.
+
+You can group regexes with a space or a pipe (`|`) between them. A space means that Alex should match each regex successively, while the pipe means that it can alternate between choices, such as `0 (x [0-9a-fA-F]+ | o [0-7]+)`, which is a regex that can match hexadecimal or octal numbers.
```

Comment:
This is a somewhat convoluted explanation of concatenation. First of all, you don't need a space, concatenating regular expressions corresponds to concatenating the languages (i.e. sets of strings) they define (Cartesian-product style). Second, whitespace in Alex regexes is simply ignored (unless it's escaped).